liberty-maven-plugin/src/main/java/net/wasdev/wlp/maven/plugins/server/DevMojo.java

```diff
@@ -0,0 +1,889 @@
+                    }
+                }
+
+                // if queue size >= 1, it means a newer test has been queued so we
+                // should skip this and let that run instead
+                if (executor.getQueue().size() >= 1) {
+                    Runnable head = executor.getQueue().peek();
+                    boolean manualInvocation = ((TestJob) head).isManualInvocation();
+
+                    if (manualInvocation) {
+                        log.info("Tests were invoked while previous tests were running. Restarting tests.");
+                    } else {
+                        log.info("Changes were detected while tests were running. Restarting tests.");
+                    }
+                    return;
+                }
+
+                if (!skipITs) {
+                    if (waitForApplicationUpdate) {
+                        // wait until application has been updated
+                        if (appUpdateTimeout < 0) {
+                            appUpdateTimeout = 5;
+                        }
+                        long timeout = appUpdateTimeout * 1000;
+                        serverTask.waitForUpdatedStringInLog(regexp, timeout, logFile, messageOccurrences);
+                    }
+
+                    log.info("Running integration tests...");
+                    try {
+                        runIntegrationTests();
+                        log.info("Integration tests finished.");
+                    } catch (MojoExecutionException e) {
+                        Throwable cause = e.getCause();
+                        if (cause != null && cause instanceof MojoFailureException) {
+                            log.debug(e);
+                            log.error("Integration tests failed: " + cause.getLocalizedMessage());
+                        } else {
+                            log.error("Failed to run integration tests", e);
+                        }
+                    }
+                }
+            }
+
+            // finally, start watching for hotkey presses if not already started
+            util.runHotkeyReaderThread(executor);
+        }
+
+        @Override
+        public void checkConfigFile(File configFile, File serverDir) {
+            try {
+                ServerFeature servUtil = new ServerFeature();
+                Set<String> features = servUtil.getServerFeatures(serverDir);
+                features.removeAll(existingFeatures);
+                if (!features.isEmpty()) {
+                    List<String> configFeatures = new ArrayList<String>(features);
+                    log.info("Configuration features have been added");
+                    runMojo("net.wasdev.wlp.maven.plugins:liberty-maven-plugin", "install-feature", serverName,
+                            configFeatures);
+                    this.existingFeatures.addAll(features);
+                }
+            } catch (Exception e) {
+                log.debug("Failed to read configuration file", e);
```

Comment:
Should we throw a `MojoExecutionException` here instead of just logging it? If something went wrong when installing features then we can't expect dev mode to continue working properly.

Reply:
This installs features when server.xml changes.  Ideally we would want dev mode to keep working if it was a development related error by the user.  When we call `runMojo` which uses `executeMojo`, `MojoExecutionException` could be thrown for a number of possible reasons, including if the mojo ran properly but features failed to install (e.g. if user specified invalid feature name).  We wouldn't want to quit dev mode in this case, because the user needs to be allowed to correct their error.  I've changed it to catch `MojoExecutionException` but log an error instead of debug.